merino-suggest/src/lib.rs

```diff
@@ -291,6 +291,31 @@ pub trait SuggestionProvider: Send + Sync {
     fn is_null(&self) -> bool {
         false
     }
+
+    /// Generate a set of cache inputs for a given query specific to this
+    /// provider. Any property of the query that affects how suggestions are
+    /// generated should be included.
+    ///
+    /// By default, all properties of the query are used, but providers should
+    /// narrow this to a smaller scope.
+    fn cache_inputs(&self, req: &SuggestionRequest, hasher: &mut blake3::Hasher) {
```

Comment:
I would have liked to keep `blake3` out of the "public" interface, but I couldn't find a way that I was happy with. Initially I tried to make this argument a closure that would take `&[u8]` that the provider could call multiple times. I couldn't get that to be both object safe for the trait and also lifetime safe.

I want to avoid `cache_inputs` just returning a `Vec` of strings or `&[u8]`s, because I think allocating multiple times for every provider for every request would get us in trouble.

Reply:
I also tried such silly things as `Arc<Box<dyn Fn(&[u8])>>` and `std::sync::mpsc::Channel`.

Reply:
No real clue on how to work around this either.

Reply:
@pjenvey do you have any clever ideas here? We can definitely ship this as is, but my JS-brain is unhappy about this, and I hoped there could be something better. :joy:

Reply:
Yeah, it's not ideal to expose this particular hasher type here. I wonder if we can make this a generic function, and then wrap the black::Hasher struct with a new type and implement the `write` method by simply making a `update` call. 

```rs
use std::hash::Hasher;

fn cache_inputs<T: Hasher>(&self, req: &SuggestionRequest, hasher: &mut T) {
   ...
}

struct Blake3Hasher(blake::Hasher);

impl Hasher for Black3Hasher {
  fn write(&mut self, bytes: &[u8]) {
    self.update(bytes);
  }
}
```

A little bit boilerplate, but it works with the standard Hasher trait, which allows us to change it in the future if needed.

Reply:
```rust
fn cache_inputs<T: Hasher>
```

We can't make `cache_inputs` have any generic parameters, since `SuggestionProvider` is used primarily as a trait object, and generics aren't object safe. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a5070587d7766b5bfe77dd6e440f795a
https://doc.rust-lang.org/book/ch17-02-trait-objects.html#trait-objects-perform-dynamic-dispatch

If we could do generics, this could use a closure directly with

```rust
fn cache_inputs<F: Fn(&[u8])>(&self, req: &SuggestionRequest, closure: F);
```

Reply:
Ah, now I know what this object safety is all about.

And we cannot do `hasher: impl Hasher` in the trait implementation; `Box<dyn Hasher>` is not cheat either. 🤷‍♂️

Reply:
`Box<dyn Hasher>` isn't too bad. I tried something similar to that with closures. I wasn't able to get the lifetimes sorted with the closure, but maybe I could with `Box<dyn Hasher>`

Reply:
Could also use just `hasher: &mut dyn Hasher` here for the same dynamic dispatch without the Box.

`blake3::Hasher` even implements `std::io::Write` which could be used for the trait here but its `write` method is a little annoying, returning an io Result.

Reply:
`Box<dyn Hasher>`, compared to `Box<Dyn Fn(&[u8])>`, has the advantage that I can add a `finish` method to it. That fixes the major problem I had with the closure, because I couldn't easily get the blake3 hasher back once I gave it to the closure.

Reply:
`&mut dyn Hasher` works great. I had forgotten that would work. I forget sometimes that you can do unsized stuff in other ways that `Box`.

Reply:
Right, dynamic dispatching works for references too.